src/sema/eval.rs

```diff
@@ -196,3 +202,398 @@ pub fn eval_const_rational(
+                } else {
+                    Ok(Expression::NumberLiteral(*loc, ty.clone(), left % right))
+                }
+            } else {
+                Ok(Expression::Modulo(
+                    *loc,
+                    ty.clone(),
+                    Box::new(left),
+                    Box::new(right),
+                ))
+            }
+        }
+        Expression::Power(loc, ty, unchecked, left, right) => {
+            let left = eval_constants_in_expression(&*left)?;
+            let right = eval_constants_in_expression(&*right)?;
+
+            if let (
+                Expression::NumberLiteral(_, _, left),
+                Expression::NumberLiteral(right_loc, _, right),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right, &Type::Uint(16), right_loc) {
+                    Err(diagnostic)
+                } else {
+                    Ok(Expression::NumberLiteral(
+                        *loc,
+                        ty.clone(),
+                        left.pow(right.to_u16().unwrap().into()),
+                    ))
+                }
+            } else {
+                Ok(Expression::Power(
+                    *loc,
+                    ty.clone(),
+                    *unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                ))
+            }
+        }
+        Expression::ShiftLeft(loc, ty, left, right) => {
+            let left = eval_constants_in_expression(&*left)?;
+            let right = eval_constants_in_expression(&*right)?;
+
+            if let (
+                Expression::NumberLiteral(_, _, left),
+                Expression::NumberLiteral(right_loc, _, right),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right, &Type::Uint(64), right_loc) {
+                    Err(diagnostic)
+                } else {
+                    Ok(Expression::NumberLiteral(
+                        *loc,
+                        ty.clone(),
+                        left.shl(right.to_u64().unwrap()),
+                    ))
+                }
+            } else {
+                Ok(Expression::ShiftLeft(
+                    *loc,
+                    ty.clone(),
+                    Box::new(left),
+                    Box::new(right),
+                ))
+            }
+        }
+
+        Expression::ShiftRight(loc, ty, left, right, bool) => {
+            let left = eval_constants_in_expression(&*left)?;
+            let right = eval_constants_in_expression(&*right)?;
+
+            if let (
+                Expression::NumberLiteral(_, _, left),
+                Expression::NumberLiteral(right_loc, _, right),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right, &Type::Uint(64), right_loc) {
+                    Err(diagnostic)
```

Comment:
This is very ugly. This means that if I do:
```solidity
int128 a = 105 >> 5192296858534827628530496329220096;
```
Now we get a diagnostic about `5192296858534827628530496329220096` not fitting into `uint64`. That's not helpful.

We need a better diagnostic in this case and also a test case for this.

Reply:
I noticed that constant folding throws a hard error there. I think the best action is to not check on the the right hand side as solc doesn't do this, it just looks as the result of the shifting

Reply:
> I can throw a warning if the shift is done with a number larger than the bit size, what do you think?

Yes, that would be a good solution. Also note the shift might be negative which is also not allowed (so the diagnostic should be "shift by negative value not permitted".

Reply:
What I fear here is when constant folding is enabled, there will be both a hard error and a warning for the same issue.
Check: https://github.com/hyperledger/solang/blob/fe342c5ae3dc6de823c8144f833b1dd522da8a78/src/codegen/constant_folding.rs#L527

Reply:
Given that, I only check for the final result of the shift

Reply:
If sema produces an error, I don't think codegen is executed so  this shouldn't be a problem.